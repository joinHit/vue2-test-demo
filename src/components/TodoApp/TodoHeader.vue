<template>
  <header class="header">
    <h1>todos</h1>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
      autofocus
      data-testid="new-todo"
      @keyup.enter="handleNewTodo"
    />
  </header>
</template>

<script>
export default {
  name: 'TodoHeader',
  methods: {
    handleNewTodo(e) {
      const value = e.target.value.trim()
      // 没有输入内容（就回车），应该被拦截
      if (value.length === 0) {
        return
      }

      this.$emit('new-todo', value)
      e.target.value = ''
    }
  }
}
</script>
