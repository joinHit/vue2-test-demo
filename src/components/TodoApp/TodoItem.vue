<template>
  <!-- <li class="completed">
      <div class="view">
        <input class="toggle" type="checkbox" checked />
        <label>Taste JavaScript</label>
        <button class="destroy"></button>
      </div>
      <input class="edit" value="Create a TodoMVC template" />
    </li> -->
  <li
    data-testid="todo-item"
    :class="{
      completed: checkboxStatus,
      editing: isEditing
    }"
  >
    <div class="view">
      <input
        class="toggle"
        type="checkbox"
        data-testid="todo-done"
        v-model="checkboxStatus"
      />
      <label data-testid="todo-text" @dblclick="isEditing = true">{{
        todo.text
      }}</label>
      <button
        class="destroy"
        data-testid="delete-todo"
        @click="$emit('delete-todo', todo.id)"
      ></button>
    </div>
    <input
      class="edit"
      data-testid="todo-edit"
      @blur="isEditing = false"
      v-focus123="isEditing"
      :value="todo.text"
      @keyup.enter="handleEditTodo"
      @keyup.esc="handleCancelEditTodo"
    />
  </li>
</template>

<script>
export default {
  name: 'TodoItem',
  props: {
    todo: {
      type: Object,
      required: true
    }
  },
  directives: {
    focus123(element, binding) {
      /** 编辑状态下，自动获得焦点
       *
       * binding.value 的值为 "isEditing 的值"
       * element 的值为 "（含有 v-focus123的）html元素"
       *
       * 解释：
       * html上，有这么一段代码 v-focus123="isEditing"
       */
      if (binding.value) {
        element.focus()
      }
    }
  },
  computed: {
    checkboxStatus: {
      // 更新 "todo项 的选中" 状态
      get() {
        return this.todo.done
      },
      // 改变 "todo项 的选中" 状态
      set(isChecked) {
        this.$emit('change-todo-done', {
          id: this.todo.id,
          done: isChecked
        })
      }
    }
  },
  data() {
    return {
      isEditing: false
    }
  },
  methods: {
    // 按下 "回车键" 时，处理 "编辑中的 todo项"
    handleEditTodo(e) {
      this.$emit('edit-todo', {
        id: this.todo.id,
        text: e.target.value
      })

      // 取消 "编辑状态"
      this.handleCancelEditTodo()
    },

    // （按下 "ESC 键"）取消 "编辑中的 todo项"
    handleCancelEditTodo() {
      this.isEditing = false
    }
  }
}
</script>
