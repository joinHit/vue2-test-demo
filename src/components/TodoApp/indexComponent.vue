<template>
  <section class="todoapp">
    <todo-header @new-todo="handleNewTodo"></todo-header>

    <!-- This section should be hidden by default and shown when there are todos -->
    <section class="main">
      <input
        id="toggle-all"
        class="toggle-all"
        type="checkbox"
        data-testid="toggle-all"
        v-model="toggleAll"
      />
      <label for="toggle-all">Mark all as complete</label>
      <ul class="todo-list">
        <!-- These are here just to show the structure of the list items -->
        <!-- List items should get the class `editing` when editing and `completed` when marked as completed -->
        <TodoItem
          v-for="todo in filterTodoList"
          :key="todo.id"
          :todo="todo"
          @delete-todo="handleDeleteTodo"
          @edit-todo="handleEditTodo"
          @change-todo-done="changeTodoDone"
        />
      </ul>
    </section>

    <!-- This footer should be hidden by default and shown when there are todos -->
    <TodoFooter :todoList="todoList" @clear-completed="handleClearCompleted" />
  </section>
</template>

<script>
import TodoHeader from './TodoHeader.vue'
import TodoFooter from './TodoFooter.vue'
import TodoItem from './TodoItem.vue'

export default {
  name: 'indexComponent',
  components: {
    TodoHeader,
    TodoFooter,
    TodoItem
  },
  data() {
    const todoList = [
      {
        id: 0,
        text: '00',
        done: false
      },
      {
        id: 1,
        text: '01',
        done: true
      },
      {
        id: 2,
        text: '02',
        done: false
      }
    ]

    return {
      // 下拉列表
      todoList,
      // 下拉列表中的 "每一项 id"
      id: todoList.length
    }
  },
  computed: {
    toggleAll: {
      get() {
        /** 设置 toggleAll 的选中状态
         *
         * 此处特别注意一下：
         *  this.todoList为空数组时，经过every方法，也会返回true。
         *  但这样就会造成 "web界面" 的 "全选 按钮" 是 "选中 状态"（不符合我们的认知）
         *     我们的认知是：有选项且全部选中时，"全选 按钮" 才 "选中"。
         *     但现在是 "没有1个 todo选项" 的情况下，"全选 按钮" 也被选中了
         *  所以需要加上 this.todoList.length 判断（是否为空数组）
         */
        return this.todoList.length && this.todoList.every((item) => item.done)
      },
      /** 点击 "全选的 复选框" 时，可以拿到 "isChecked参数
       * 该参数代表：全选框，是否被选中
       */
      set(isChecked) {
        this.todoList.forEach((item) => {
          item.done = isChecked
        })
      }
    },

    // （根据路由）过滤数据
    filterTodoList() {
      // 获取路由路径
      const path = this.$route.path

      // 根据路由路径，过滤数据
      // 路径 /           代表 "（todo列表上的）所有todo"
      // 路径 /active     代表 "（todo列表上的）所有未完成的todo"
      // 路径 /completed  代表 "（todo列表上的）所有已完成的todo"
      switch (path) {
        case '/':
          return this.todoList
        case '/active':
          return this.todoList.filter((item) => !item.done)
        case '/completed':
          return this.todoList.filter((item) => item.done)
        default:
          return this.todoList
      }
    }
  },
  methods: {
    // 添加 todo列表
    handleNewTodo(text) {
      this.todoList.push({
        id: this.id,
        text,
        done: false
      })

      this.id++
    },

    // 删除某个 todo
    handleDeleteTodo(todoId) {
      this.todoList = this.todoList.filter((item) => item.id !== todoId)
    },

    // 按下 "回车键" 时，处理 "编辑中的 todo项"
    handleEditTodo(todo) {
      const { id, text } = todo
      const findTodo = this.todoList.find((item) => item.id === id)

      if (!findTodo) {
        return
      }

      // 执行删除操作
      if (text.trim().length === 0) {
        this.handleDeleteTodo(id)
      }

      // 执行修改操作
      findTodo.text = text
    },

    // 改变 "todo项 的选中" 状态
    changeTodoDone(todo) {
      const findTodo = this.todoList.find((item) => item.id === todo.id)

      if (findTodo) {
        findTodo.done = todo.done
      }
    },

    // 清除 "选中的 clear completed" 的 "todo项"
    handleClearCompleted() {
      this.todoList = this.todoList.filter((item) => !item.done)
    }
  }
}
</script>
