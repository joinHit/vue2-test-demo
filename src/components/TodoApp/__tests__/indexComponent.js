import TodoHeader from '@/components/TodoApp/TodoHeader.vue'
import TodoItem from '@/components/TodoApp/TodoItem.vue'
import indexComponent from '@/components/TodoApp/indexComponent.vue'
import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'

describe('indexComponent.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper = null

  // 模拟的列表数据
  const todoList1 = [
    {
      id: 0,
      text: 'eat',
      done: false
    },
    {
      id: 1,
      text: 'play',
      done: true
    },
    {
      id: 2,
      text: 'sleep',
      done: false
    }
  ]

  beforeEach(async () => {
    // 定义 "当前页面的路由"
    const $route = {
      path: '/'
    }

    wrapper = shallowMount(indexComponent, {
      mocks: {
        $route
      }
    })

    // 设置 todoList 的值（方法1）
    /*     wrapper.vm.todoList = todoList1
    await Vue.nextTick() */

    // 设置 todoList 的值（方法2）
    await wrapper.setData({
      todoList: JSON.parse(JSON.stringify(todoList1)),
      id: todoList1.length
    })
  })

  // 调用当前组件的方法
  test('handle new todo', () => {
    // 设置 "子组件（TodoHeader.vue）" 传过来的值
    const text = 'play'

    // 调用当前组件的方法
    wrapper.vm.handleNewTodo(text)

    const todo = wrapper.vm.todoList.find((item) => item.text === text)

    expect(todo).toBeTruthy()
  })

  // 添加 "列表数据"，更新 "列表视图"
  it('todo list', () => {
    expect(wrapper.findAllComponents(TodoItem).length).toBe(todoList1.length)
  })

  // 模拟调用子组件的方法，更新视图列表(相当于 "前面 2个测试用例")
  test('new todo', async () => {
    // 设置 "子组件（TodoHeader.vue）" 传过来的值
    const text = 'play'

    // 模拟 2次 调用子组件的方法
    await wrapper.getComponent(TodoHeader).vm.$emit('new-todo', text)
    await wrapper.getComponent(TodoHeader).vm.$emit('new-todo', text)

    expect(wrapper.findAllComponents(TodoItem).length).toBe(
      todoList1.length + 2
    )
  })

  // 删除 "某个 todo" 成功
  test('delete todo success', async () => {
    // 存在 "todo 的 id 为 1" 的测试（真实存在的测试，也叫 "正向测试"）
    await wrapper.vm.handleDeleteTodo(1)

    // 因为总共有 "3 个数据"，又删除了id为 "1 的数据"，所以还剩余 "2 个数据"
    expect(wrapper.vm.todoList.length).toBe(2)
    expect(wrapper.findAllComponents(TodoItem).length).toBe(2)
  })

  // 删除 "某个 todo" 失败
  test('delete todo fail', async () => {
    // 不存在 "todo 的 id 为 123" 的测试（根本不存在的测试，也叫 "反向测试"）
    await wrapper.vm.handleDeleteTodo(123)

    // 因为总共有 "3 个数据"，又不存在id为 "123 的数据"，所以还剩余 "3 个数据"
    expect(wrapper.vm.todoList.length).toBe(3)
    expect(wrapper.findAllComponents(TodoItem).length).toBe(3)
  })

  // 编辑 "某个 todo"
  test('edit todo', async () => {
    // （异步）从 "todo列表" 中 获取 单项todo
    async function getTodoItem(todo) {
      await wrapper.vm.handleEditTodo(todo)

      return wrapper.vm.todoList.find((item) => item.id === todo.id)
    }

    // （TodoItem.vue）子组件传来的参数
    const emitTodo = {
      id: 2,
      text: 'abc'
    }

    // 正向测试（输入的是：有效数据）
    expect((await getTodoItem(emitTodo))?.text).toBe(emitTodo.text)

    // 反向测试（输入的是：无效数据。即空字符串）
    // （TodoItem.vue）子组件传来的 参数text：置空（代表删除该 item项）
    emitTodo.text = ''
    // expect(await getTodoItem(emitTodo)).toBeFalsy()
    expect(await getTodoItem(emitTodo)).toBeUndefined()
  })

  // 点击 "全选（todo）" 按钮
  test('click toggle all', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')

    // 1、"全部 todo" 选中的情况
    // 相当于: 点击 "全选" 按钮，使其为 "选中" 状态
    await toggleAll.setChecked()
    // 断言 "所有的（todo）子任务" 都被选中了
    wrapper.vm.todoList.forEach((item) => {
      expect(item.done).toBeTruthy()
    })

    // 2、"全部 todo" 未选中的情况
    // 相当于: 点击 "全选" 按钮，使其为 "未选中" 状态
    await toggleAll.setChecked(false)
    // 断言 "所有的（todo）子任务" 都被选中了
    wrapper.vm.todoList.forEach((item) => {
      expect(item.done).toBeFalsy()
    })
  })

  // 点击 "（todoList）列表前的（每一个）选项" 按钮
  test('click item all', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')

    // 1、让所有任务都变成完成状态，
    // 即："每一项 todo 前" 的选项按钮，都要点击一下，让其选上
    wrapper.vm.todoList.forEach((item) => {
      item.done = true
    })
    // （"全选按钮" 的状态变化）需要等待视图更新
    await Vue.nextTick()

    // 断言 "全选按钮" 也选中了
    expect(toggleAll.element.checked).toBeTruthy()

    // 2、取消摸一个任务（置为未完成状态）时，断言 "全选按钮" 未完成
    // 此处为：把 "todo列表中的第0项" 数据的状态，置为 "未完成 状态"
    wrapper.vm.todoList[0].done = false
    await Vue.nextTick()
    expect(toggleAll.element.checked).toBeFalsy()
  })

  // 清除 "已完成（代指：已选中）" 的 "todo项"
  test('clear completed todo ', async () => {
    wrapper.vm.handleClearCompleted()
    await Vue.nextTick()

    const todoList2 = todoList1.filter((item) => !item.done)
    expect(wrapper.vm.todoList).toEqual(todoList2)
  })

  // （根据路由）过滤数据
  test('filter todoList', async () => {
    // 1、 将路由导航到 /
    wrapper.vm.$route.path = '/'
    await Vue.nextTick()
    // 断言 filterTodoList  = （todo列表上的）所有todo
    expect(wrapper.vm.filterTodoList).toEqual(todoList1)

    // 2、 将路由导航到 /active
    wrapper.vm.$route.path = '/active'
    await Vue.nextTick()
    // 断言 filterTodoList  = （todo列表上的）所有未完成的todo"
    const todoListActive = todoList1.filter((item) => !item.done)
    expect(wrapper.vm.filterTodoList).toEqual(todoListActive)

    // 3、 将路由导航到 /completed
    wrapper.vm.$route.path = '/completed'
    await Vue.nextTick()
    // 断言 filterTodoList  = （todo列表上的）所有已完成的todo"
    const todoListCompleted = todoList1.filter((item) => item.done)
    expect(wrapper.vm.filterTodoList).toEqual(todoListCompleted)
  })

  // 快照
  test('snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
