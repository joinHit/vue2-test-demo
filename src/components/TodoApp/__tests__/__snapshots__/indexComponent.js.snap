// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexComponent.vue snapshot 1`] = `
<section class="todoapp">
  <todo-header-stub></todo-header-stub>
  <section class="main"><input id="toggle-all" type="checkbox" data-testid="toggle-all" class="toggle-all"> <label for="toggle-all">Mark all as complete</label>
    <ul class="todo-list">
      <todoitem-stub todo="[object Object]"></todoitem-stub>
      <todoitem-stub todo="[object Object]"></todoitem-stub>
      <todoitem-stub todo="[object Object]"></todoitem-stub>
    </ul>
  </section>
  <todofooter-stub todolist="[object Object],[object Object],[object Object]"></todofooter-stub>
</section>
`;
